doc/man/lcmaps_pilot_sub_proxy.mod.8.in
.TH LCMAPS_PILOT_SUB_PROXY.MOD 8 "May 6, 2015" "@PACKAGE_NAME@ @VERSION@"                                                                           
.SH NAME
lcmaps_pilot_sub_proxy.mod \- LCMAPS plugin to handle pilot sub-proxies
.SH SYNOPSIS
.nh
.ad l
.B lcmaps_pilot_sub_proxy.mod
.RB [ \-\-add-pilot-fqans
.IR yes | no ]
.RB [ \-\-require-limited
.IR yes | no ]
.RB [ \-\-match-fqan
.IR pattern ]
.RB [ \-\-lock-type
.IR none | flock | fcntl ]
.RB [ \-\-max-pcpathlen
.IR value ]
.SH DESCRIPTION
This plugin is meant to be used in a very specific pilot job scenario, where the
payload user has no certificate of its own, but the pilot reliably knows the
identity of the payload user.

In order for this plugin to run, the pilot should create one more proxy
delegation encoding the username into the extra CN field of the proxy. This new
payload proxy should be passed into the LCMAPS framework as the normal PEM
string (e.g. for gLExec via the GLEXEC_CLIENT_CERT environment variable). The
original pilot proxy should be pointed to by the X509_USER_PROXY.

The plugin will verify (in this order) that
.\" use TP instead of groff specific nr
.TP 3
1.
the pilot proxy contains its private key,
.TP
2.
both the pilot and payload proxy are RFC3820 type proxies,
.TP
3.
both the pilot and payload proxy are limited proxies (unless
\fB\-\-require-limited\fR is set to \fIno\fR),
.TP
4.
at least one of the FQANs matches a specified pattern (when using
\fB\-\-match-fqan\fR),
.TP
5.
the payload proxy is signed by the pilot proxy,
.TP
6.
the effective proxy pathlength contraint for the payload proxy does not exceed
the maximum (normally 0, unless changed with \fB\-\-max-pcpathlen\fR),
.TP
7.
the value of the extra proxy commonName RDN (CN) does not start with a
slash (/).
.PP
If all checks succeed, the plugin will add the value of the extra payload proxy
commonName RDN into the set of LCMAPS credentials as the user_dn. Additionally
it will also add the FQANs of the original proxy to this set of LCMAPS
credentials unless the flag \fB\-\-add-pilot-fqans\fR is set to \fIno\fR.

.SH OPTIONS
.TP
.BI "\-\-add-pilot-fqans "{yes|no}
In addition to adding the subject DN of the leaf-payload-proxy, the plugin can
also add the FQANs of the proxy. Default is \fIyes\fR, to add the FQANs.

.TP
.BI "\-\-require-limited "{yes|no}
Specifies whether an extra test is done to check that both proxies are limited.
Default is \fIyes\fR, to do this test.

.TP
.BI "\-\-match-fqan "FQAN-pattern
When specified, at least one of the FQANs needs to match the given pattern. It
is a wildcard supporting matching. E.g. '*/Role=pilot*'. Note that this option
has limited usability unless the client is doing the VOMS verification.

.TP
.BI "\-\-lock-type "{none|flock|fcntl}
Type of locking mechanism used for reading in the pilot proxy pointed to by the
X509_USER_PROXY, default is \fInone\fR.

.TP
.BI "\-\-max-pcpathlen "maximum\ effective\ pcPathLen
Maximum effective proxy pathlength constraint for the leaf payload proxy.
Default is \fI0\fR, i.e. no proxies are allowed after it. Use \fI-1\fR for
infinite. \fBNOTE\fR: under normal circumstances this option should be left to
the default value 0.

.SH RETURN VALUES
.TP
.B LCMAPS_MOD_SUCCESS
Success.
.TP
.B LCMAPS_MOD_FAIL
Failure.

.SH ENVIRONMENT
.TP
X509_USER_PROXY
should point to a file containing the proxy of the pilot job user. It should be
a valid RFC3820 compliant proxy.

.SH EXAMPLE
The following example config file can be used for LCMAPS:
.PP  
.br
 # default path for the modules
.br
 path = <libdir>/lcmaps
.br

.br
 # Plugin definitions:
.br
 verifyproxy = "lcmaps_verify_proxy.mod"
.br
               " \-certdir /etc/grid\-security/certificates"
.br

.br
 pilot_sub_proxy = "lcmaps_pilot_sub_proxy.mod"
                   " \-\-match-fqan */Role=pilot*"
.br

.br
 scasclient = "lcmaps_scas_client.mod"
.br
              " \-capath /etc/grid-security/certificates/"
.br
              " \-endpoint https://scas.example.org:8443"
.br
              " \-resourcetype wn"
.br
              " \-actiontype execute-now"
.br
              " \-\-use-dn-and-fqans"
.br

.br
 # Policies:
.br
 pilot_sub_proxy:
.br
 verifyproxy \-> pilot_sub_proxy
.br
 pilot_sub_proxy \-> scasclient
.br


.RE

.SH NOTES
.IP (1) 4
Since this plugin only checks that the payload proxy is signed by the leaf-proxy
of the pilot proxy-chain, it is very important to run first the
lcmaps_verify_proxy.mod which will verify the latter proxy chain.
.IP (2)
Since any user allowed to call gLExec could provide these type of proxies, it is
necessary to verify that the pilot proxy at least has the pilot role.
.IP (3)
When running inside gLExec (typical scenario), it is most probably necessary to
prevent the payload from obtaining the proxy. This can be enforced by the pilot:
export GLEXEC_TARGET_PROXY=/dev/null

.P
A typical invocation in gLExec would be something like
.P
\fC $ create_pilot_sub_proxy.sh User:JohnDoe\fR
.br
\fC $ export GLEXEC_CLIENT_CERT=${X509_USER_PROXY}_payload\fR
.br
\fC $ export GLEXEC_TARGET_PROXY=/dev/null\fR

.SH SEE ALSO
.BR lcmaps.db (5), 
.BR lcmaps (3),
.BR lcmaps_plugins_scas_client (8),
.BR lcmaps_verify_proxy.mod (8),
.BR glexec (1)

.SH AUTHORS
LCMAPS and the LCMAPS plug-ins were written by the Grid Middleware Security Team
<grid-mw-security@nikhef.nl>.


NEWS
Version 0.0.1:
- initial release of lcmaps-plugins-pilot-sub-proxy: plugin to put the
  'pilot-sub-proxy' username into the LCMAPS credential data.


src/lcmaps_pilot_sub_proxy.c
/**
 * Copyright (c) FOM-Nikhef 2015-
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *
 *  Authors:
 *  2015-
 *     Mischa Sall\'e <msalle@nikhef.nl>
 *     NIKHEF Amsterdam, the Netherlands
 *     <grid-mw-security@nikhef.nl>
 *
 */

/**
 * NOTES: lcmaps-verify-proxy checks payload chain->if one longer than pilot:
 * pilot also checked. Only check whether is indeed only one longer. So check
 * cert by cert until bottom of stack. Probably don't need to rebase on a CA.
 * Still need external check that pilot may do this: on GUMS can check pilot DN
 * is allowed to make pilot proxies (has role pilot) */

#include <openssl/x509.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include "lcmaps_plugins_pilot_sub_proxy_config.h"

#include <lcmaps/lcmaps_defines.h>
#include <lcmaps/lcmaps_arguments.h>
#include <lcmaps/lcmaps_log.h>

#if defined(HAVE_LCMAPS_LCMAPS_PLUGIN_PROTOTYPES_H)
#   include <lcmaps/lcmaps_plugin_prototypes.h>
#else
#   include "lcmaps_plugin_prototypes.h"
#endif

#include "lcmaps_pilot_sub_proxy_utils.h"


/************************************************************************
 * defines
 ************************************************************************/

#define PLUGIN_PREFIX	"lcmaps_pilot_sub_proxy"

#define PLUGIN_RUN	0   /* full run mode */
#define PLUGIN_VERIFY	1   /* verify-only mode */


/************************************************************************
 * global variables
 ************************************************************************/

/** Whether to only register DN or also FQANs, default yes */
static int add_pilot_fqans=1;

/** Whether to require proxies are limited, default yes */
static int require_limited=1;

/** When set, the FQAN pattern must match one of the FQANs */
static char *fqan_pattern=NULL;

/** Lock type to use for reading the X509_USER_PROXY */
static lock_type_t lock_type=LOCK_NOLOCK;

/** Default maximum effective proxy pathlen for leaf payload proxy */
static long maxpcpathlen=0;


/************************************************************************
 * private prototypes
 ************************************************************************/

/* called by plugin_run() and plugin_verify() */
static int plugin_run_or_verify(int argc, lcmaps_argument_t *argv,
				int lcmaps_mode);


/************************************************************************
 * public functions
 ************************************************************************/

/**
 * Initialize function for plugin
 */
int plugin_initialize(int argc, char **argv) {
    const char * logstr = PLUGIN_PREFIX"-plugin_initialize()";
    char *endptr=NULL;
    int i;

    /* Log commandline parameters on debug */
    lcmaps_log(LOG_DEBUG,"%s: passed arguments:\n",logstr);
    for (i=0; i < argc; i++)
	lcmaps_log(LOG_DEBUG,"%s: arg %d is %s\n", logstr, i, argv[i]);

    /* Parse arguments, argv[0] = name of plugin, so start with i = 1 */
    for (i = 1; i < argc; i++) {
	if (strcmp(argv[i], "--add-pilot-fqans") == 0)
	{
	    if (argv[i + 1] == NULL)	{
		lcmaps_log(LOG_ERR,
		    "%s: option %s needs to be followed by 'yes' or 'no'\n",
		    logstr, argv[i]);
		return LCMAPS_MOD_FAIL;
	    }
	    if (strcmp(argv[i+1],"yes") == 0)  {
		lcmaps_log(LOG_DEBUG,
		    "%s: will add FQANs from pilot when available\n", logstr);
		add_pilot_fqans=1;
	    } else if (strcmp(argv[i+1],"no") == 0)  {
		lcmaps_log(LOG_DEBUG,
		    "%s: will NOT add FQANs from pilot\n", logstr);
		add_pilot_fqans=0;
	    } else {
		lcmaps_log(LOG_ERR,
		    "%s: option %s should have value 'yes' or 'no', not '%s'\n",
		    logstr, argv[i], argv[i]);
		return LCMAPS_MOD_FAIL;
	    }
	    i++;
	}
	else if (strcmp(argv[i], "--require-limited") == 0)
	{
	    if (argv[i + 1] == NULL)	{
		lcmaps_log(LOG_ERR,
		    "%s: option %s needs to be followed by 'yes' or 'no'\n",
		    logstr, argv[i]);
		return LCMAPS_MOD_FAIL;
	    }
	    if (strcmp(argv[i+1],"yes") == 0)  {
		lcmaps_log(LOG_DEBUG,
		    "%s: require proxies to be limited\n", logstr);
		require_limited=1;
	    } else if (strcmp(argv[i+1],"no") == 0)  {
		lcmaps_log(LOG_DEBUG,
		    "%s: do NOT require proxies to be limited\n", logstr);
		require_limited=0;
	    } else {
		lcmaps_log(LOG_ERR,
		    "%s: option %s should have value 'yes' or 'no', not '%s'\n",
		    logstr, argv[i], argv[i]);
		return LCMAPS_MOD_FAIL;
	    }
	    i++;
	}
	else if (strcmp(argv[i], "--match-fqan") == 0)
	{
	    if (argv[i + 1] == NULL)	{
		lcmaps_log(LOG_ERR,
		    "%s: option %s needs to be followed by FQAN pattern\n",
		    logstr, argv[i]);
		return LCMAPS_MOD_FAIL;
	    }
	    if (argv[i + 1][0]!='\0')
		fqan_pattern=argv[i + 1];
	    i++;
	}
	else if (strcmp(argv[i], "--lock-type") == 0)
	{
	    if (argv[i + 1] == NULL)	{
		lcmaps_log(LOG_ERR,
		    "%s: option %s needs to be followed by valid lock type\n",
		    logstr, argv[i]);
		return LCMAPS_MOD_FAIL;
	    }
	    if (strcmp(argv[i+1], "none") == 0)	{
		lcmaps_log(LOG_INFO,
			"%s: not using locking for reading X509_USER_PROXY\n",
			logstr);
		lock_type=LOCK_NOLOCK;
	    } else if (strcmp(argv[i+1], "fcntl") == 0) {
		lcmaps_log(LOG_INFO,
			"%s: using fcntl locking for reading X509_USER_PROXY\n",
			logstr);
		lock_type=LOCK_FCNTL;
	    } else if (strcmp(argv[i+1], "flock") == 0)	{
		lcmaps_log(LOG_INFO,
			"%s: using flock locking for reading X509_USER_PROXY\n",
			logstr);
		lock_type=LOCK_FLOCK;
	    } else    {
		lcmaps_log(LOG_ERR, "%s: unknown lock_type \"%s\"\n",
			logstr, argv[i+1]);
		return LCMAPS_MOD_FAIL;
	    }
	    i++;
	}
	else if (strcmp(argv[i], "--max-pcpathlen") == 0)
	{
	    /* check valid argument */
	    if (argv[i + 1] == NULL || argv[i + 1][0]=='\0') {
		lcmaps_log(LOG_ERR,
		    "%s: option %s needs to be followed by a valid number\n",
		    logstr, argv[i]);
		return LCMAPS_MOD_FAIL;
	    }

	    /* try to convert argument to long */
	    errno=0;
	    maxpcpathlen=strtol(argv[i + 1],&endptr,10);
	    /* endptr will either point to first non-valid char, or end of valid
	     * string unless argv[i+1]=="", which we already checked above */
	    if ( errno!=0 || *endptr!='\0' ||
		 (maxpcpathlen<0 && maxpcpathlen!=-1) )
	    {
		lcmaps_log(LOG_ERR,
		    "%s: Illegal value for \"%s\" (%s): "
		    "Should be a (long int)>=0 or -1\n",
		    logstr, argv[i], argv[i + 1]);
		return LCMAPS_MOD_FAIL;
	    }
            i++;
	}
	else
	{
            lcmaps_log(LOG_ERR,
		    "%s: Unknown argument for plugin: %s (failure)\n",
		    logstr, argv[i]);
            return LCMAPS_MOD_FAIL;
        }
    }

    return LCMAPS_MOD_SUCCESS;
}

/**
 * Introspect function for plugin, defines which lcmaps framework data is made
 * available as commandline options for the plugin_run.
 */
int plugin_introspect(int *argc, lcmaps_argument_t **argv) {
    const char * logstr = PLUGIN_PREFIX"-plugin_introspect()";
    static lcmaps_argument_t argList[] = {
	{"user_dn"    ,"char *"          , 1,NULL},
	{"nfqan"      ,"int"             , 0,NULL},
	{"fqan_list"  ,"char **"         , 0,NULL},
	{"px509_chain","STACK_OF(X509) *", 0,NULL},
	{"pem_string" ,"char *"          , 0,NULL},
	{NULL         ,NULL              ,-1,NULL}
    };

    lcmaps_log(LOG_DEBUG,"%s: introspecting\n", logstr);

    *argv = argList;
    *argc = lcmaps_cntArgs(argList);
    lcmaps_log(LOG_DEBUG,"%s: address first argument: %p\n",
	    logstr, (void*)argList);

    return LCMAPS_MOD_SUCCESS;
}

/**
 * run function for plugin, wrapper around plugin_run_or_verify() with
 * PLUGIN_RUN mode
 */
int plugin_run(int argc, lcmaps_argument_t *argv) {
    return plugin_run_or_verify(argc, argv, PLUGIN_RUN);
}

/**
 * run function for plugin, wrapper around plugin_run_or_verify() with
 * PLUGIN_VERIFY mode
 */
int plugin_verify(int argc, lcmaps_argument_t *argv) {
    return plugin_run_or_verify(argc, argv, PLUGIN_VERIFY);
}

/**
 * terminate plugin, frees used memory.
 * return LCMAPS_MOD_SUCCESS
 */
int plugin_terminate(void) {
    const char * logstr = PLUGIN_PREFIX"-plugin_terminate()";

    lcmaps_log(LOG_DEBUG,"%s: terminating\n", logstr);

    return LCMAPS_MOD_SUCCESS;
}



/************************************************************************
 * private functions
 ************************************************************************/

/**
 * Actual run/verify function. Called by both plugin_run and plugin_verify
 * with different lcmaps_mode.
 */
static int plugin_run_or_verify(int argc, lcmaps_argument_t *argv,
				int lcmaps_mode) {
    const char *        logstr       = NULL;
    STACK_OF(X509)*	pilot_chain  = NULL;
    STACK_OF(X509)*	payload_chain= NULL;
    X509 *		pilot_cert   = NULL;
    X509 *		payload_cert = NULL;
    int                 nfqans       = -1;
    char **             fqans        = NULL;
    long                pcpathlen;

    /* Set suitable logstr */
    if (lcmaps_mode == PLUGIN_RUN)
        logstr = PLUGIN_PREFIX"-plugin_run()";
    else if (lcmaps_mode == PLUGIN_VERIFY)
        logstr = PLUGIN_PREFIX"-plugin_verify()";
    else {
        lcmaps_log(LOG_ERR, PLUGIN_PREFIX"-plugin_run_or_verify(): "
		"attempt to run plugin in invalid mode: %d\n", lcmaps_mode);
        goto fail_plugin;
    }

    /* Get X509_USER_PROXY */
    if (psp_get_pilot_proxy(&pilot_chain, lock_type))
	goto fail_plugin;

    /* Get payload proxy (typically PEM string)	*/
    if (psp_get_payload_proxy(&payload_chain, argc, argv))
	goto fail_plugin;

    /* Get FQANs when needed */
    if (add_pilot_fqans || fqan_pattern)    {
	if (psp_get_fqans(&nfqans, &fqans, argc, argv))
	    goto fail_plugin;
    }

    /* Get leaf proxies */
    pilot_cert=sk_X509_value(pilot_chain,0);
    payload_cert=sk_X509_value(payload_chain,0);
    if (pilot_cert==NULL || payload_cert==NULL)	{
	lcmaps_log(LOG_WARNING, "%s: cannot get leaf proxy certs from chains\n",
		logstr);
	goto fail_plugin;
    }

    /* Check whether chains are valid RFC proxies */
    if (psp_proxy_is_rfc(pilot_cert)==0)    {
	lcmaps_log(LOG_WARNING,
	    "%s: pilot proxy is not RFC compliant\n", logstr);
	goto fail_plugin;
    }
    if (psp_proxy_is_rfc(payload_cert)==0)	{
	lcmaps_log(LOG_WARNING,
	    "%s: payload proxy is not RFC compliant\n", logstr);
	goto fail_plugin;
    }

    /* Check whether either proxy is not LIMITED */
    if (require_limited)    {
	if (psp_proxy_is_limited(pilot_cert)==0)	{
	    lcmaps_log(LOG_WARNING,
		"%s: pilot proxy is not a Limited proxy\n", logstr);
	    goto fail_plugin;
	}
	if (psp_proxy_is_limited(payload_cert)==0)	{
	    lcmaps_log(LOG_WARNING,
		"%s: payload proxy is not a Limited proxy\n", logstr);
	    goto fail_plugin;
	}
    }

    /* Check pattern */
    if (fqan_pattern && (psp_match_fqan(nfqans, fqans, fqan_pattern)==0))   {
	lcmaps_log(LOG_WARNING,
	    "%s: proxy does not contain required FQAN(-pattern) %s\n",
	    logstr, fqan_pattern);
	goto fail_plugin;
    }

    /* Do actual verification */
    if (psp_verify_proxy_signature(payload_cert, pilot_cert))
	goto fail_plugin;
    
    /* Get effective pcPathLen for payload proxy, this also checks we have not
     * already exceeded it */
    if (psp_get_pcpathlen(payload_chain, &pcpathlen))
	goto fail_plugin;

    /* Check that the remaining effective proxy pathlength constraint for the
     * payload proxy is not too large */
    if ( maxpcpathlen>=0 && (pcpathlen==-1 || pcpathlen>maxpcpathlen) )	{
	lcmaps_log(LOG_WARNING,
		"%s: effective proxy pathlength constraint for payload proxy "
		"is too large: %ld (maximum: %ld)\n",
		__func__, pcpathlen, maxpcpathlen);
	goto fail_plugin;
    }

    /* After doing last check, whether it doesn't start with a slash, store the
     * value of the last /CN=... of the payload cert as user_dn */
    if (psp_store_proxy_dn(payload_cert, pilot_cert))
	goto fail_plugin;

    /* Store the FQANs of the proxy when add_pilot_fqans==1 */
    if (add_pilot_fqans && psp_store_fqans(nfqans, fqans))
	goto fail_plugin;
   
    /* Cleanup chain memory */
    psp_cleanup_chains(pilot_chain, payload_chain);

    lcmaps_log(LOG_INFO,"%s: %s plugin succeeded\n", logstr, PLUGIN_PREFIX);

    return LCMAPS_MOD_SUCCESS;

fail_plugin:
    /* Cleanup chain memory */
    psp_cleanup_chains(pilot_chain, payload_chain);

    lcmaps_log(LOG_INFO,"%s: %s plugin failed\n", logstr, PLUGIN_PREFIX);

    return LCMAPS_MOD_FAIL;
}
